### ISSUE TITLE
Router should not export Type Interface

### ISSUE THREAD

**Issue by naomiblack** [angular/angular#3632]
We currently get two docs exported for Type Interface: one for core, and one for router. This is confusing -- and I think unnecessary duplication? Can we make Router depend on the Type exported from core?

**Comment by btford** [comment]
The router does depend on `Type` from core, but because it exposes a class with member of type `Type`, it must also re-export `Type` or TypeScript is unhappy.

@mprobst @alexeagle any ideas?

**Comment by mprobst** [comment]
I think that is WAI in TypeScript, if you don't re-export, downstream clients cannot understand your type signatures. You could re-export or relax the field to be of type `any`.

is the field private by chance? Because if so, then it'd be a valid feature request for TS to support this use case (not being required to re-export private fields).

**Comment by btford** [comment]
The field needs to be public.

I think this is something the documentation system is going to have to
change to handle.

**Comment by alexeagle** [comment]
Does this result in an error for users who import both typings? @jeffbcross checked that it should work.
If it's just documented twice, how severe an issue is that?

**Comment by btford** [comment]
This should be innocuous for the user. I think `type Type = Function` or something similar.

We could extend dgeni to filter out certain duplicates. I suspect we'll encounter the same problem with animations, http, etc. It might be reasonable to solve the general issue of duplicate exports. Or we could hard code a work around for now.

I think this will only really happen with shared utils like the façades. Could we check if the given class originates from `src/facade/*` and just not export those definitions?

**Comment by btford** [comment]
@naomiblack – I'm going to unassign myself since this is not actionable on my end.

**Comment by naomiblack** [comment]
Assigning to Misko to think about best fix.

**Comment by mhevery** [comment]
After some research here is what needs to be done:

1. Don't export `Type` from `router.ts`
2. include `///<reference path="./angular2.d.ts"/>` in the `router.d.ts` file. This can be done by updating `/Users/misko/work/angular/docs/typescript-definition-package/index.js` to include `./angular2.d.ts`
3. Have dgeni reffer to type as `ng.Type` rather then `Type`

I know how do do step 1 and 2, (see this #3704) no idea how to do step 3. Re-asigning to @btford , as I am hoping that he knows how to do 3, if not reasign to @petebacondarwin.

**Comment by btford** [comment]
I have some idea how to get dgeni to do this.

Are there other cases where we need to refer to a class with an `ng.`-prefix?

**Comment by mprobst** [comment]
@mhevery I think we should avoid `/// <reference path=...>` syntax. It should be ok for the router to depend on Angular - after all it does. `/// <reference>` is odd syntax and doesn't fit with our use of ES6 modules.

Can't we make dgeni smarter to handle this situation - re-exporting an aliased type from a different package? Seems like something our doc generator should handle. This is probably just a matter of resolving aliased symbols, TypeScript has an explicit API to do that:
```ts
https://github.com/angular/ts2dart/blob/master/lib/facade_converter.ts#L141
```

**Comment by mhevery** [comment]
@mprobst Type is exported from `angular2.d.ts`. Router depends on `angualr2.d.ts` and it uses its `Type`. How do you propose we solve this situation?

This is not a `dgeni` issue. What is a canonical way for `two.d.ts` to use types from `one.d.ts`?

**Comment by mprobst** [comment]
`.d.ts` vs `.ts` should not make a difference except for the `declare` modifier. The `.d.ts` files should look like this:

a.d.ts:
```ts
declare module "whatever/a" {
  export class Wobble {}
}
```

b.d.ts
```ts
declare module "whatever/b" {
  import {Wobble} from "whatever/a";
  export var w: Wobble;
}
```

**Comment by alexeagle** [comment]
Misko and I tried that, but the compiler still complains that this mixes
ambient declarations with a module. I think the import keyword is causing
b.d.ts to be a module, even though it should not have any emit for that
statement.
Seems similar to a couple other bugs I've filed where Mohammed has relaxed
the constraints in an ambient context, where the expression won't emit
anything, for example https://github.com/Microsoft/TypeScript/issues/2929
Misko, are you still experimenting with this?
